### ISSUE TITLE
front does not work with join

### ISSUE THREAD

**Issue by thopper21** [ericniebler/range-v3#1070]
I want to use `ranges::front` to get the first element in a joined range, like so:

```
std::vector<int> is = ...;
auto foo = front(
    is | view::transform([](auto& i) { return view::repeat_n("foo", i); }) |
    view::join);
```

This feels like it should work, but it does not compile.

**Comment by thopper21** [comment]
Interestingly enough, this appears to work with `action::join`. I'm still surprised that it does not work with the `view` flavor.

**Comment by CaseyCarter** [comment]
The `join` result here is an `InputRange`, but not a `ForwardRange`. `front` rejects non-forward ranges to make it harder to write bugs like:
```c++
auto r = /**/;
front(r);
begin(r);
```
with single-pass ranges. (Yes, this is a common mistake to make with single-pass ranges: I found several occurrences in range-v3 itself when I made `front` and `empty` reject non-forward ranges.)

`*begin(r)` is equivalent to `front(r)`, and just enough harder to write that people have to pause and think about what they're doing. This is by design.
